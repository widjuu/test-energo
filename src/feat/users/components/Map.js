import { MapContainer, TileLayer, Marker, Popup } from "react-leaflet";
import { useSelector } from "react-redux";

export const Map = () => {
  const selectedUsers = useSelector((state) => state.selectedUsers.selected);

  console.log(selectedUsers);

  const markers = Object.keys(selectedUsers).map((userId) => {
    const user = selectedUsers[userId];

    if (!user) {
      return null;
    }

    const lat = user?.address?.geo?.lat ?? 0;
    const lng = user?.address?.geo?.lng ?? 0;
    const name = user?.name ?? "Без имени";

    return (
      <Marker position={[lat, lng]} key={user.id}>
        <Popup>{name}</Popup>
      </Marker>
    );
  });

  console.log(markers);

  return (
    <MapContainer
      center={[0, 0]}
      zoom={1}
      scrollWheelZoom={false}
      className="leaflet"
    >
      <TileLayer
        attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
        url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
      />
      {markers}
    </MapContainer>
  );
};
