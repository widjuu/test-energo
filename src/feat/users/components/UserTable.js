import { useDispatch, useSelector } from "react-redux";

//components
import { Table } from "antd";

//utils
import { toggleUser } from "../actions";

const columns = [
  {
    title: "Имя",
    dataIndex: "name",
  },
  {
    title: "Телефон",
    dataIndex: "phone",
  },
];

export const UserTable = () => {
  const dispatch = useDispatch();
  const data = useSelector((state) => state.users.data);
  const selectedUsers = useSelector((state) => state.selectedUsers.selected);

  const dataSource = data.map((userId) => {
    return {
      id: userId?.id,
      name: userId?.name,
      key: userId?.id,
      phone: userId?.phone,
      address: userId?.address,
    };
  });

  const selectedRowKeys = Object.keys(selectedUsers).map((userId) => {
    const user = selectedUsers[userId];

    return user.id;
  });

  const rowSelection = {
    selectedRowKeys,
    onSelect(record) {
      dispatch(
        toggleUser({
          id: record.id,
          zipcode: record.zipcode,
          address: record.address,
        })
      );
    },
  };

  return (
    <div>
      <Table
        size="small"
        pagination={false}
        columns={columns}
        dataSource={dataSource}
        rowSelection={rowSelection}
      />
    </div>
  );
};
