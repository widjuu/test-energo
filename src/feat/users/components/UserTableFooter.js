import { useSelector, useDispatch } from "react-redux";

// components
import { makeStyles } from "@material-ui/core/styles";
import Box from "@material-ui/core/Box";
import Button from "@material-ui/core/Button";
import CircularProgress from "@material-ui/core/CircularProgress";

// redux
import { sendSelectedUsers } from "../actions";

const useStyles = makeStyles(({ palette }) => ({
  error: {
    color: palette.error.main,
  },
}));

export const UserTableFooter = () => {
  const classes = useStyles();
  const dispatch = useDispatch();

  const selectedUsers = useSelector((state) => state.selectedUsers.selected);
  const loading = useSelector((state) => state.selectedUsers.loading);
  const error = useSelector((state) => state.selectedUsers.error);

  const disabled = Object.keys(selectedUsers).length === 0;

  const onButtonClick = () => dispatch(sendSelectedUsers(selectedUsers));

  return (
    <Box display="flex" alignItems="center" px={3} my={1}>
      <Button
        startIcon={
          loading ? <CircularProgress size={20} color="secondary" /> : undefined
        }
        onClick={onButtonClick}
        disabled={disabled}
        variant="contained"
        color={error ? "secondary" : "primary"}
      >
        Отправить
      </Button>
      {error ? (
        <Box ml={2} className={classes.error}>
          {error}
        </Box>
      ) : null}
    </Box>
  );
};
