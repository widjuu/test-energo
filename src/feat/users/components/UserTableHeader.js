import TableHead from "@material-ui/core/TableHead";
import TableCell from "@material-ui/core/TableCell";
import TableRow from "@material-ui/core/TableRow";

export const UserTableHeader = () => {
  return (
    <TableHead>
      <TableRow>
        <TableCell>Имя</TableCell>
        <TableCell align="right">Телефон</TableCell>
        <TableCell padding="checkbox"></TableCell>
      </TableRow>
    </TableHead>
  );
};
