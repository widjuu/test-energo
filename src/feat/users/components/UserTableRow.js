import { useDispatch, useSelector } from "react-redux";

// components
import TableCell from "@material-ui/core/TableCell";
import TableRow from "@material-ui/core/TableRow";
import Checkbox from "@material-ui/core/Checkbox";

import { toggleUser } from "../actions";

export const UserTableRow = ({ id, name, phone, address }) => {
  const dispatch = useDispatch();
  const selectedUsers = useSelector((state) => state.selectedUsers.selected);
  const isSelected = !!selectedUsers[id];

  const onClick = () => dispatch(toggleUser({ id, zipcode: address?.zipcode }));

  return (
    <TableRow hover>
      <TableCell>{name}</TableCell>
      <TableCell align="right">{phone}</TableCell>
      <TableCell padding="checkbox">
        <Checkbox checked={isSelected} onChange={onClick} />
      </TableCell>
    </TableRow>
  );
};
