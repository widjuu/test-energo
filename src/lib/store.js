import { configureStore } from "@reduxjs/toolkit";
import logger from "redux-logger";

import api from "./api";
import { usersReducer, selectedUsersReducer } from "../feat/users/reducer";

const isDev = process.env.NODE_ENV !== "production";

const store = configureStore({
  reducer: {
    users: usersReducer.reducer,
    selectedUsers: selectedUsersReducer.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      thunk: {
        extraArgument: { api },
      },
    }).concat(isDev ? [logger] : []),
  devTools: isDev,
});

export default store;
