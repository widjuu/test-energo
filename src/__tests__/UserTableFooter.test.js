import React from "react";
import renderer from "react-test-renderer";
import { render, unmountComponentAtNode } from "react-dom";

import { act } from "react-dom/test-utils";
import { UserTableFooter } from "../feat/users/components/UserTableFooter";

import { Provider } from "react-redux";
import store from "../lib/store";

let container = null;
beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it("test button", () => {
  act(() => {
    render(
      <Provider store={store}>
        <UserTableFooter />
      </Provider>,
      container
    );
  });
  expect(container.querySelector("Button").disabled).toBeDefined();
  expect(container.textContent).toBe("Отправить");
});

describe("UserTableFooter Snapshot", () => {
  it("+++capturing UserTableFooter Snapshot", () => {
    const renderedValue = renderer
      .create(
        <Provider store={store}>
          <UserTableFooter />
        </Provider>
      )
      .toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
});
