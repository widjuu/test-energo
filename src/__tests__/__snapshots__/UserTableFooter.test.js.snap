// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserTableFooter Snapshot +++capturing UserTableFooter Snapshot 1`] = `
<div
  className="user-table-footer"
>
  <button
    className="ant-btn ant-btn-primary ant-btn-round"
    color="primary"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    <span>
      Отправить
    </span>
  </button>
</div>
`;
